### PR TITLE
xds/resolver: Add support for cluster specifier plugins

### PR DESCRIPTION
**Pull request by zasweq**
This PR adds support for cluster specifier specifier plugins in the xds resolver/config selector, as per RLS in xDS design. The functionality still left to implement is the RLS Cluster Specifier Plugin.

RELEASE NOTES: None

### CODE REVIEW

File: xds/internal/resolver/serviceconfig.go
Comment (dfawley) [inline xds/internal/resolver/serviceconfig.go:138]: Can we put the CSP name into clusters instead so we don't need another field?

File: xds/internal/resolver/watch_service.go
Comment (dfawley) [inline xds/internal/resolver/watch_service.go:167]: If this is the right place to put this, then I think the function needs a new name.

File: xds/internal/resolver/serviceconfig.go
Comment (zasweq) [reply xds/internal/resolver/serviceconfig.go:138]: I see what your saying (persist as little state as possible), but I much prefer keeping this as a separate field. Adding csp to clusters would conflate an individual csp name to a WRR type which can hold onto many cluster names and chooses one randomly. This keeps the branching logic (two fields in state, two separate ways of persistence/handling), which started from the xdsclient.

File: xds/internal/resolver/watch_service.go
Comment (zasweq) [reply xds/internal/resolver/watch_service.go:167]: I think this is the right place to put it (mechanical conversion from xdsclient struct -> xds resolver representation of data). Do you see any place else? The vh encapsulates the specific WRR cluster/csp name per route, and this is just the top level. The best name I could think of (after a few minutes of thinking haha) is updateStateFromRDS. Let me know if there's a better naming choice.

**Review by zasweq — COMMENTED** [review]
Thanks for the comments :D!

File: xds/internal/resolver/serviceconfig.go
Comment (dfawley) [reply xds/internal/resolver/serviceconfig.go:138]: It's not about "persisting state", it's about reducing & simplifying code and data structures.  We use the same code paths already for when WRR is used vs. when a single cluster is specified directly, so why not fold CSP into it as well?  In gRPC, they are equivalent: CSPs are the same as ordinary clusters from the name resolver and cluster manager LB policy's perspectives.

File: xds/internal/resolver/watch_service.go
Comment (dfawley) [reply xds/internal/resolver/watch_service.go:167]: I agree this is where it belongs.  It seems like this is the _real_ "handle RDS" function, and `handleRDSResp` should not do anything besides call this (and handle errors first), which is what it does.  Maybe rename to `applyRouteConfigUpdate`?

File: xds/internal/resolver/serviceconfig.go
Comment (zasweq) [reply xds/internal/resolver/serviceconfig.go:138]: Kept the branching in newConfigSelector (for prefix + hardcoded weight 1 logic, which was in unmarshal_rds for singular cluster)...but now it's one codepath in SelectConfig/one piece of state clusters.

File: xds/internal/resolver/watch_service.go
Comment (zasweq) [reply xds/internal/resolver/watch_service.go:167]: Renamed to applyRouteConfigUpdate.

**Review by zasweq — COMMENTED** [review]
Thanks for the comments! I like the simplicity push. Easwar mentioned you're good at finding simple solutions to problems.

File: xds/internal/resolver/serviceconfig.go
Comment (dfawley) [inline xds/internal/resolver/serviceconfig.go:377]: Shouldn't this be `"cluster_specifier_plugin:" + rt.ClusterSpecifierPlugin`?

Please use a local to hold that instead of repeating it 4 times.  Also, global consts for `"cluster:"` and `"cluster_specifier_plugin:"` would be a good idea to avoid any chance of a typo in one usage.

File: xds/internal/resolver/serviceconfig.go
Comment (dfawley) [inline xds/internal/resolver/serviceconfig.go]: Can this bit be shared with the below section?  Maybe factored into another function?

File: xds/internal/resolver/serviceconfig.go
Comment (zasweq) [reply xds/internal/resolver/serviceconfig.go:377]: Yeah, good catch. Added global consts and also a local var for both branches.

Comment (zasweq) [reply xds/internal/resolver/serviceconfig.go:377]: Also, the comment isn't showing up on this PR, but I refactored the cluster initialization into another function. It made it much cleaner.

**Review by zasweq — COMMENTED** [review]
Thanks for the comments :D!

File: xds/internal/resolver/serviceconfig.go
Comment (dfawley) [inline xds/internal/resolver/serviceconfig.go:95]: So... _if_ a CSP is called "cluster:foo" _and_ a cluster named "foo" exists, then "cluster:foo" will be found in the CSP map, even though that isn't what was intended.

Should we be looking at whether the `cluster` from `activeClusters`'s key `StartsWith` `clusterSpecifierPluginPrefix` instead here?

File: xds/internal/resolver/serviceconfig.go
Comment (dfawley) [inline xds/internal/resolver/serviceconfig.go:96]: Are we handling this case correctly (I think we're not):

1. First xDS update contains a cluster specifier plugin for RLS
2. RPC starts that selects that CSP
3. New xDS update removes that CSP
4. When generating this JSON here, we need to maintain the CSP config data and keep the "cluster" it references around until (2) is done

File: xds/internal/balancer/cdsbalancer/cdsbalancer.go
Comment (zasweq) [reply xds/internal/balancer/cdsbalancer/cdsbalancer.go:184]: Why would it be better to strip in the cluster manager lb policy? To me, it makes the most sense to strip it here, as the addition of the cluster: prefix was for cluster type cds. Stripping it that early leads me to think there would be name collisions somewhere (either in router or in cluster manager itself). Wdyt @menghanl?

Comment (zasweq) [reply xds/internal/balancer/cdsbalancer/cdsbalancer.go:184]: Discussed in person, decided it would be best it to strip this in the name resolver, while constructing the Cluster Manager LB Config itself. (kept the cluster: prefix for the name of the cluster, but stripped it for the underlying cds policy)

File: xds/internal/resolver/serviceconfig.go
Comment (zasweq) [reply xds/internal/resolver/serviceconfig.go:95]: I agree that there is a correctness issue here, but I disagree that this issue can be fixed by what you are describing. What about if a cluster has the prefix "cluster_specifier_plugin:foo", starts with would return true, and could have a configuration in clusterSpecifierPlugins keyed on foo.

File: xds/internal/resolver/serviceconfig.go
Comment (zasweq) [reply xds/internal/resolver/serviceconfig.go:96]: Ah, I found a solution that solves this problem and the comment about name collisions on the branch in serviceConfigJSON. I simply added the csp balancer configuration (if the cluster is a csp) to the cluster info. This reuses all the plumbing around active clusters, and also solves the correctness issue on the branch of the cluster specifier, as you can branch on clusterInfo.cspCfg != nil.

File: xds/internal/resolver/serviceconfig.go
Comment (zasweq) [reply xds/internal/resolver/serviceconfig.go:95]: Simply persisted the config in clusterInfo, which solves this correctness issue.

**Review by zasweq — COMMENTED** [review]
Thanks for the comments :D! Adding test case for the situation outlined by you.

**Comment by dfawley** [comment]
As mentioned offline, this is blocked on a decision of how to handle CSP config changes.

File: xds/internal/resolver/serviceconfig.go
Comment (dfawley) [inline xds/internal/resolver/serviceconfig.go:415]: I think this wants to be a method on `configSelector` instead.  It can access the `xdsResolver` that way, too for adding to `activeClusters`.  You can also pass it a `balancerConfig` to store in the `clusterInfo` as mentioned above.

File: xds/internal/resolver/watch_service.go
Comment (dfawley) [inline xds/internal/resolver/watch_service.go:39]: nit: contain**S** (the `...Plugins` map is a singular)

File: xds/internal/resolver/xds_resolver_test.go
Comment (dfawley) [inline xds/internal/resolver/xds_resolver_test.go:868]: Can you move this & related tests to a new file?  This one is already too big.

Maybe `cluster_specifier_plugin_test.go`?

File: xds/internal/resolver/xds_resolver_test.go
Comment (dfawley) [inline xds/internal/resolver/xds_resolver_test.go:886]: A quick 1-liner description of what this tests would be helpful.  E.g. `// Tests that cluster specifier plugins produce the correct service config.` or something SG.

File: xds/internal/resolver/xds_resolver_test.go
Comment (dfawley) [inline xds/internal/resolver/xds_resolver_test.go:1343]: Please add another test that covers updates to a CSP configuration - it should change in the output service config.

File: xds/internal/resolver/xds_resolver_test.go
Comment (dfawley) [inline xds/internal/resolver/xds_resolver_test.go:1291]: The formatting here looks slightly off.

File: xds/internal/resolver/xds_resolver_test.go
Comment (dfawley) [inline xds/internal/resolver/xds_resolver_test.go:1330]: Formatting again

File: xds/internal/resolver/serviceconfig.go
Comment (zasweq) [reply xds/internal/resolver/serviceconfig.go:415]: Ah, great idea. These two comments make it much cleaner. Switched.

**Review by zasweq — COMMENTED** [review]
Thanks for the comments :D!

File: xds/internal/resolver/cluster_specifier_plugin_test.go
Comment (dfawley) [inline xds/internal/resolver/cluster_specifier_plugin_test.go:317]: New comment from old comment thread, since the code was moved:

This should line up with "cluster_specifier_plugin:cspA" (outdent 1 tab).  And the same situation below @ line 356.

It's not a correctness thing, it's just formatting.

**Review by dfawley — APPROVED** [review]
LGTM modulo the formatting nit.

**Pull request merged**
